src/java/org/apache/cassandra/db/compaction/unified/Controller.java

```diff
@@ -855,10 +865,12 @@ public static Controller fromOptions(CompactionRealm realm, Map<String, String>
                                                   ? Reservations.Type.valueOf(options.get(RESERVATIONS_TYPE_OPTION).toUpperCase())
                                                   : DEFAULT_RESERVED_THREADS_TYPE;
 
-        if (options.containsKey(NUM_SHARDS_OPTION))
+        if (options.containsKey(NUM_SHARDS_OPTION) || DEFAULT_NUM_SHARDS.isPresent())
         {
             // Legacy V1 mode.
-            int numShards = Integer.parseInt(options.get(NUM_SHARDS_OPTION));
+            int numShards = options.containsKey(NUM_SHARDS_OPTION) // options take precedence
```

Comment:
Assuming that users would be gradually upgraded to density sharding, and that we would likely want to switch some problem cases early as well as test, shouldn't we provide a way of cancelling the defined default `num_shards` via the table definition?

Reply:
Interesting idea. I was thinking about gradually removing the settings kube by kube. Do you think this would be too coarse grained?

Reply:
I opted for a special value for the existing NUM_SHARDS property rather than a new property. 
If V1 compatibility mode is set via system property, NUM_SHARDS table property may be set to `-1` value to switch to V2.